pkg/gconfig/gconfig.go

```diff
@@ -0,0 +1,232 @@
+package gconfig
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+)
+
+// Config is the list of variables which a provider can be configured with.
+type Config []*Field
+
+// Get a config value. Useful for testing purposes.
+// If the config value is secret a redacted string will be returned.
+// Returns an empty string if not found.
+func (c Config) Get(key string) string {
+	for _, v := range c {
+		if v.Key() == key {
+			return v.String()
+		}
+	}
+	return ""
+}
+
+type Dumper interface {
+	Dump(ctx context.Context, cfg Config) (map[string]string, error)
+}
+
+// Loader loads configuration for Granted providers.
+type Loader interface {
+	// Load configuration. Returns a map of config values.
+	// The keys of the map are the value names, and the values
+	// are the actual values of the config.
+	// For example:
+	//	{"orgUrl": "http://my-org.com"}
+	//
+	// Returns an error if loading the values fails.
+	Load(ctx context.Context) (map[string]string, error)
+}
+
+// Load configuration using a loader.
+func (c Config) Load(ctx context.Context, l Loader) error {
+	loaded, err := l.Load(ctx)
+	if err != nil {
+		return err
+	}
+	for _, s := range c {
+		key := s.Key()
+		val, ok := loaded[key]
+		if !ok && !s.IsOptional() {
+			return fmt.Errorf("could not find %s in map", key)
+		}
+		s.Set(val)
+	}
+	return nil
+}
+
+// Dump renders a map[string]string where the values are mapped in different ways based on the provided dumper
+//
+// use SafeDumper to get all values with secrets redacted
+//
+// SSMDumper first pushes any updated secrets to ssm then returns the ssm paths to the secrets
+func (c Config) Dump(ctx context.Context, dumper Dumper) (map[string]string, error) {
```

Comment:
Do you think we need the `dumper` interface? I'm wondering if it might be simpler to just have different structs with a `Dump` method e.g.

```go
var c Config
s := SSMDumper{}
s.Dump(ctx, c)
```

rather than the current approach:
```go
var c Config
s := SSMDumper{}
c.Dump(ctx, s)
```

My only criticism of the current approach is that it feels a bit less intuitive as you'd expect the SSMDumper to 'do' the dumping, (`s.Dump()`) rather than the config to do the dumping.

We might also have this problem with the `config.Load()` method which I originally wrote!

Reply:
I think I will just make the safeDump and SSMDump methods on Config directly

Reply:
interesting point. I think its important for "easy to use, hard to misuse" that if you have a config, its obvious how you get all the values out.

Our specific use-case requires SSM to be involved. 

As a developer using this library i think, "i have a Config, how do I get the values out"
I use autocomplete to list the methods available then I read the description of Dump.
If we removed dump there would be no autocomplete :(
I propose we leave it in for now and can factor it out later if its not clear enough.